src/Tronkell/Network/Websockets.hs

```diff
@@ -0,0 +1,67 @@
+module Tronkell.Network.Websockets where
+
+import qualified Network.WebSockets as WS
+import qualified Network.Wai.Handler.Warp as Warp
+import qualified Network.Wai.Handler.WebSockets as WaiWS
+import qualified Network.Wai.Application.Static as Static
+import qualified Network.Wai
+import qualified Control.Concurrent as Con
+import qualified Data.Aeson as A
+import Data.ByteString.Lazy.Char8
+import Control.Exception (handle, SomeException(..))
+import Control.Monad (forever, when)
+import Debug.Trace
+
+import qualified Tronkell.Server.Types as ST
+import qualified Tronkell.Data.Parse as TP
+
+start :: Con.MVar ST.UserID -> ST.NetworkChans -> Con.Chan ST.OutMessage -> IO ()
+start uIdGen chans outChan = Warp.runSettings
+  (Warp.setPort 8331 Warp.defaultSettings)
+  $ WaiWS.websocketsOr WS.defaultConnectionOptions (websocketHandler uIdGen chans outChan) staticApp
+
+staticApp :: Network.Wai.Application
+staticApp = Static.staticApp . Static.defaultWebAppSettings $ "ui/public"
+
+websocketHandler :: Con.MVar ST.UserID -> ST.NetworkChans -> Con.Chan ST.OutMessage -> WS.ServerApp
+websocketHandler uIdGen (inChan, clientSpecificOutChan) outChan pendingConnection = do
+  conn <- WS.acceptRequest pendingConnection
+  WS.forkPingThread conn 30
+  -- duplicate the channels to read.
+  dupOutChan <- Con.dupChan outChan
+  dupClientSpecificOutChan <- Con.dupChan clientSpecificOutChan
+
+  -- get from some mvar passed from server to here.
+  userId <- nextUserID uIdGen
```

Comment:
Why not do this in the server? Because we'd need one more message to tell the user what their id would be?